### PR TITLE
make intern_const_alloc_recursive return error

### LINKED ISSUE #78655: ICE: compiler/rustc_middle/src/mir/mod.rs:2306:49: could not find allocation for alloc1
### Code

```Rust
const FOO: *const u32 = {
    let x;
    &x
};
fn main() {
    let FOO = FOO;
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.49.0-nightly (4f7612ac1 2020-10-31)
binary: rustc
commit-hash: 4f7612ac1499258025077f1fd05d2f429f9accfb
commit-date: 2020-10-31
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
```

### Error output

```
error[E0381]: borrow of possibly-uninitialized variable: `x`
 --> reduced_mutant.rs:3:5
  |
3 |     &x
  |     ^^ use of possibly-uninitialized `x`

error: encountered dangling pointer in final constant
 --> reduced_mutant.rs:1:1
  |
1 | / const FOO: *const u32 = {
2 | |     let x;
3 | |     &x
4 | | };
  | |__^

error[E0005]: refutable pattern in local binding: `_` not covered
 --> reduced_mutant.rs:6:9
  |
1 | / const FOO: *const u32 = {
2 | |     let x;
3 | |     &x
4 | | };
  | |__- constant defined here
5 |   fn main() {
6 |       let FOO = FOO;
  |           ^^^
  |           |
  |           interpreted as a constant pattern, not a new variable
  |           help: introduce a variable instead: `foo_var`
  |
  = note: the matched value is of type `*const u32`

error: internal compiler error: compiler/rustc_middle/src/mir/mod.rs:2306:49: could not find allocation for alloc1

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:942:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.49.0-nightly (4f7612ac1 2020-10-31) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_borrowck] borrow-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 4 previous errors

Some errors have detailed explanations: E0005, E0381.
For more information about an error, try `rustc --explain E0005`.
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:942:9
stack backtrace:
   0: std::panicking::begin_panic
   1: rustc_errors::HandlerInner::bug
   2: rustc_errors::Handler::bug
   3: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
   4: rustc_middle::ty::context::tls::with_opt::{{closure}}
   5: rustc_middle::ty::context::tls::with_opt
   6: rustc_middle::util::bug::opt_span_bug_fmt
   7: rustc_middle::util::bug::bug_fmt
   8: rustc_middle::mir::Constant::check_static_ptr
   9: <rustc_mir::borrow_check::type_check::TypeVerifier as rustc_middle::mir::visit::Visitor>::visit_constant
  10: rustc_middle::mir::visit::Visitor::super_rvalue
  11: <rustc_mir::borrow_check::type_check::TypeVerifier as rustc_middle::mir::visit::Visitor>::visit_body
  12: rustc_mir::borrow_check::type_check::type_check
  13: rustc_mir::borrow_check::nll::compute_regions
  14: rustc_mir::borrow_check::do_mir_borrowck
  15: rustc_infer::infer::InferCtxtBuilder::enter
  16: rustc_mir::borrow_check::mir_borrowck
  17: core::ops::function::FnOnce::call_once
  18: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::mir_borrowck>::compute
  19: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  20: rustc_data_structures::stack::ensure_sufficient_stack
  21: rustc_query_system::query::plumbing::get_query_impl
  22: rustc_query_system::query::plumbing::ensure_query_impl
  23: rustc_interface::passes::analysis
  24: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute
  25: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  26: rustc_data_structures::stack::ensure_sufficient_stack
  27: rustc_query_system::query::plumbing::get_query_impl
  28: rustc_interface::passes::QueryContext::enter
  29: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  30: rustc_span::with_source_map
  31: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</p>
</details>

**NOTE: The bug is found by our work-in-progress compiler testing tool Kira, and the test program is reduced/minimized by [Perses](https://github.com/chengniansun/perses)**

### PR DESCRIPTION
**Pull request by vn-ki**
fix #78655

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_middle/src/mir/interpret/error.rs [hunk 1]
```diff
@@ -129,7 +135,7 @@ impl fmt::Display for InvalidProgramInfo<'_> {
         match self {
             TooGeneric => write!(f, "encountered overly generic constant"),
             ReferencedConstant => write!(f, "referenced constant has errors"),
-            TypeckError(ErrorReported) => {
+            AlreadyReported(ErrorReported) => {
                 write!(f, "encountered constants with type errors, stopping evaluation")
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/error.rs:139]: The error message text does not match the new variant name.

File: compiler/rustc_mir/src/interpret/intern.rs [hunk 2]
```diff
@@ -285,11 +286,13 @@ pub enum InternKind {
 /// tracks where in the value we are and thus can show much better error messages.
 /// Any errors here would anyway be turned into `const_err` lints, whereas validation failures
 /// are hard errors.
+#[tracing::instrument(skip(ecx))]
 pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
     ecx: &mut InterpCx<'mir, 'tcx, M>,
     intern_kind: InternKind,
     ret: MPlaceTy<'tcx>,
-) where
+) -> Result<(), ErrorReported>
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/intern.rs:294]: Hm, I had quite deliberately made interning infallible a while ago.^^ Is this really necessary or just yet another hack to work around https://github.com/rust-lang/rust/issues/76064 ?

Comment (oli-obk) [reply compiler/rustc_mir/src/interpret/intern.rs:294]: well... infallible or not, it was reporting errors, so it wasn't really infallible, it just didn't report its failures.

The problem isn't #76064 in this case, even if the repro test would also be fixed by a fix for #76064 . This issue also occurs for

```rust
const FOO: *const u32 = {
    let x = 42;
    &x
};
```

which typechecks just fine. Basically any dangling pointer was causing this.

As you noted in https://github.com/rust-lang/rust/issues/78655#issuecomment-720403662 that error should stop compilation. We could change the interning error to a `delay_span_bug` and have it get reported by validation instead, but then validation needs to have a look at relocations within padding, which it currently does not.

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/intern.rs:294]: > well... infallible or not, it was reporting errors, so it wasn't really infallible, it just didn't report its failures.

That's fair.

> Basically any dangling pointer was causing this.

It was causing an error, but usually not an ICE -- IIRC we even had a testcase for that.

But I guess what I really care about is that interning does not return an `InterpError`, and that is still the case. Actually telling the outside world whether we did `span_err` seems reasonable and seems to be a more general pattern.

So, LGTM.

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/intern.rs:294]: >  then validation needs to have a look at relocations within padding, which it currently does not.

Interning is in a much better place to do this, so unless we really want the error to point to where in the value the problem occurs, I think the current approach works better.

**Pull request merged**